File: code_samples/javascript_samples/sendGuidanceLines.js
const axios = require('axios');
const fs = require('fs');
const FormData = require('form-data');
const generateAccessToken = require('./generateAccessToken');

// Function to create a guidance line in Trimble Ag Software
const createGuidanceLine = async (orgId, guidanceLineData, accessToken) => {
    try {
        const headers = {
            Authorization: `Bearer ${accessToken}`,
            'Content-Type': 'application/json'
        };
        const response = await axios.post(`https://cloud.api.trimble.com/Trimble-Ag-Software/externalApi/3.0/resources/${orgId}/guidancelines/`, guidanceLineData, { headers: headers });
        console.log(response.data);
    } catch (error) {
        if (error.isAxiosError) {
            console.error(`AxiosError: ${error.message}`);
            if (error.response) {
                console.error(`Response data: ${JSON.stringify(error.response.data)}`);
            }
        } else {
            console.error(`Error: ${error}`);
        }
    }
};

// Immediately-invoked async function
(async () => {
    const accessToken = await generateAccessToken();
    let orgId;

    try {
        const orgsResponse = await axios.get('https://cloud.api.trimble.com/Trimble-Ag-Software/externalApi/3.0/config/organizations', {
            headers: {
                Authorization: `Bearer ${accessToken}`
            }
        });
        console.log(orgsResponse.data);
        orgId = orgsResponse.data.result[0].id;
    } catch (error) {
        console.error(`Error in organization fetch: ${error}`);
    }

    try {
        // Using the provided example for guidance line data
        const guidanceLineData = {
            fieldId: "189543f0-26e0-4553-8603-6dd54355da1f",
            name: "ExampleGuidanceLine",
            type: "AB",
            origin: {
                type: "Point",
                coordinates: [-104.99404, 39.73924]  // Example: Denver, CO
            },
            infillId: "fe5bc049-e206-4443-aa31-8a03772c9c62",
            geometry: {
                type: "Polygon",
                coordinates: [
                    [
                        [-104.99404, 39.73924],  // Point 1: Denver, CO
                        [-104.99404, 39.83924],  // Point 2: North of Denver, CO
                        [-105.09404, 39.83924],  // Point 3: North-west of Denver, CO
                        [-105.09404, 39.73924],  // Point 4: West of Denver, CO
                        [-104.99404, 39.73924]   // Point 5: Closing the polygon by repeating Point 1
                    ]
                ]
            },
            numberOfSwaths: 0
        };
        
        await createGuidanceLine(orgId, guidanceLineData, accessToken);
    } catch (error) {
        console.error(`Error in creating guidance line: ${error}`);
        if (error.response) {
            console.error(`Response data: ${JSON.stringify(error.response.data)}`);
        }
    }
})();

File: code_samples/python/sendGuidanceLines.py
import requests
import json
from generateAccessToken import generate_access_token

def create_guidance_line(org_id, guidance_line_data, access_token):
    try:
        headers = {
            'Authorization': f'Bearer {access_token}',
            'Content-Type': 'application/json'
        }
        response = requests.post(f'https://cloud.api.trimble.com/Trimble-Ag-Software/externalApi/3.0/resources/{org_id}/guidancelines/', json=guidance_line_data, headers=headers)
        print(response.json())
    except requests.RequestException as error:
        print(f"AxiosError: {error}")
        if error.response:
            print(f"Response data: {json.dumps(error.response.json())}")
    except Exception as error:
        print(f"Error: {error}")

def main():
    access_token = generate_access_token()
    org_id = None

    # Fetch organizations
    try:
        headers = {
            'Authorization': f'Bearer {access_token}'
        }
        orgs_response = requests.get('https://cloud.api.trimble.com/Trimble-Ag-Software/externalApi/3.0/config/organizations', headers=headers)
        orgs_data = orgs_response.json()
        print(orgs_data)
        
        org_id = orgs_data['result'][0]['id']
    except Exception as error:
        print(f"Error in organization fetch: {error}")

    # Create guidance line
    try:
        guidance_line_data = {
            'fieldId': "189543f0-26e0-4553-8603-6dd54355da1f",
            'name': "ExampleGuidanceLine",
            'type': "AB",
            'origin': {
                'type': "Point",
                'coordinates': [-104.99404, 39.73924]  # Example: Denver, CO
            },
            'infillId': "fe5bc049-e206-4443-aa31-8a03772c9c62",
            'geometry': {
                'type': "Polygon",
                'coordinates': [
                    [
                        [-104.99404, 39.73924],  # Point 1: Denver, CO
                        [-104.99404, 39.83924],  # Point 2: North of Denver, CO
                        [-105.09404, 39.83924],  # Point 3: North-west of Denver, CO
                        [-105.09404, 39.73924],  # Point 4: West of Denver, CO
                        [-104.99404, 39.73924]   # Point 5: Closing the polygon by repeating Point 1
                    ]
                ]
            },
            'numberOfSwaths': 0
        }
        
        create_guidance_line(org_id, guidance_line_data, access_token)
    except Exception as error:
        print(f"Error in creating guidance line: {error}")
        if hasattr(error, 'response') and error.response:
            print(f"Response data: {json.dumps(error.response.json())}")

if __name__ == "__main__":
    main()
